State.py

```diff
@@ -1,119 +1,108 @@
-    print("Send VoteResult ACK to ", end="")
     voteack = protoc.VoteResult()
     voteack.toAddr = toAddr
     voteack.toPort = toPort
     voteack.term = termNumber
     voteack.granted = True
-    self.server.talk(protoc.VOTERESULT, voteack)
+    return protoc.VOTERESULT, voteack
 
-  def replyAENACK(self, toAddr, toPort, termNumber):
-    print("Send AppendReply NACK to ", end="")
-    message = protoc.AppendReply()
-    message.toAddr = toAddr
-    message.toPort = toPort
-    message.term = termNumber
-    message.success = False
-    self.server.talk(protoc.APPENDREPLY, message)
-
-  def replyAEACK(self, toAddr, toPort, termNumber):
-    print("Send AppendReply ACK to ", end="")
-    message = protoc.AppendReply()
-    message.toAddr = toAddr
-    message.toPort = toPort
-    message.term = termNumber
-    message.success = True
-    self.server.talk(protoc.APPENDREPLY, message)   
 
 class LeaderState(State):
-  def __init__(self, termNumber, server):
-    State.__init__(self, termNumber, server)
+  def __init__(self):
+    State.__init__(self)
+    # Send Heartbeats once transitioned to inform a leader is present
     self.sendHeartbeat()
-    self.heartbeat = 2  # interval between heartbeat messages (this must be less than election timout lower bound)
-    #self.initTimer()
+    # Time interval to send messages in seconds
+    self.heartbeat = 4
+    self.timer = None
   
   def initTimer(self):
-    self.timer = threading.Timer(4, self.sendHeartbeat)
+    self.timer = threading.Timer(self.heartbeat, self.sendHeartbeat)
     self.timer.start()
 
   def sendHeartbeat(self):
-    #print("Sending HeartBeats")
     message = protoc.AppendEntries()
-    self.server.talk(protoc.APPENDENTRIES, message)
     self.initTimer()
+    return protoc.APPENDENTRIES, message
 
+  # TO-DO
   def handleMessage(self, messageType, message, termNumber):
-    print("")
-
-  def stop(self):
     pass
 
+
 class CandidateState(State):
-  def __init__(self, termNumber, server):
-    State.__init__(self, termNumber, server)
-    self.votes = 1   # init this to 1 because each candidate votes for themselves
+  def __init__(self):
+    State.__init__(self)
+    # Candidate has vote for himself
+    self.votes = 1  
     self.heardFromLeader = False
     self.requestVotes()
 
-  def stop(self):
-    pass
-
   def handleMessage(self, messageType, message, termNumber):
     if messageType == protoc.REQUESTVOTE:
-      self.sendVoteNACK(message.fromAddr, message.fromPort, termNumber)
+      return self.sendVoteNACK(message.fromAddr, message.fromPort, termNumber)
+    elif messageType == protoc.VOTERESULT:
+      if message.granted:
+        self.votes += 1
     elif messageType == protoc.APPENDENTRIES:
       if message.term < termNumber:
-        self.replyAENACK(message.fromAddr, message.fromPort, termNumber)
+        return self.replyAENACK(message.fromAddr, message.fromPort, termNumber)
       else:
+        # When will this ever happen?
```

Comment:
This is a situation that the Server class will handle.  I put the logic for handling the global rules (see Rules for Servers -> All Servers section on page 4 of the paper) in there because if one of those situations happens we don't care about message content.  So in short, the code will never hit this case.  It can be removed.